pageserver/src/tenant/timeline.rs

```diff
@@ -903,13 +912,43 @@ impl Timeline {
         let gc_lock = self.layer_removal_cs.lock().await;
         let mut layers = self.layers.write().unwrap();
         let mut updates = layers.batch_update();
-        self.delete_historic_layer(&gc_lock, local_layer, &mut updates)?;
-        updates.insert_historic(new_remote_layer);
+
+        let replaced = match updates.replace_historic(&local_layer, new_remote_layer)? {
```

Comment:
What do you think, should there be a comment about the `?`? That'd be a precondition failure once more, which I don't think we even be able to have -- basically meaning that right hand side lines 892..910 would be buggy. It's not an assertion because we have taken locks.

Reply:
Franking speaking I have to completely rewrite this replacement logic in  #3348.
If we are really going to somehow address the problem with storage ballooning caused with excessive image layers generation (#2948) then lets first do something with this PR before doing some refactoring.

Reply:
Oki, did not see this work colliding... I'll need to look into your PR. Please note this is not just refactoring or prettying up, it's about making the replacement across locks precise.

Reply:
@knizhnik Note we've been worried about a race condition in relation to this from #3387, which I was planning to work on next after this PR: #3533.